crates/starknet/src/starknet/add_deploy_account_transaction.rs

```diff
@@ -0,0 +1,175 @@
+use starknet_in_rust::core::errors::state_errors::StateError;
+use starknet_types::contract_address::ContractAddress;
+use starknet_types::felt::{Felt, TransactionHash};
+use starknet_types::traits::HashProducer;
+
+use super::Starknet;
+use crate::error::{Error, Result};
+use crate::traits::StateExtractor;
+use crate::transactions::deploy_account_transaction::DeployAccountTransaction;
+use crate::transactions::{StarknetTransaction, Transaction};
+
+pub fn add_deploy_account_transaction(
+    starknet: &mut Starknet,
+    deploy_account_transaction: DeployAccountTransaction,
+) -> Result<(TransactionHash, ContractAddress)> {
+    if !starknet
+        .state
+        .is_contract_declared(&Felt::new(*deploy_account_transaction.0.class_hash())?)?
+    {
+        return Err(Error::StateError(StateError::MissingClassHash()));
+    }
+
+    let state_before_txn = starknet.state.pending_state.clone();
+    let transaction_hash = deploy_account_transaction.generate_hash()?;
+    let address: ContractAddress =
+        (deploy_account_transaction.0.contract_address().clone()).try_into()?;
+
+    match deploy_account_transaction
+        .0
+        .execute(&mut starknet.state.pending_state, &starknet.block_context)
+    {
+        Ok(tx_info) => {
+            starknet.handle_successful_transaction(
+                &transaction_hash,
+                Transaction::DeployAccount(Box::new(deploy_account_transaction)),
+                tx_info,
+            )?;
+        }
+        Err(tx_err) => {
+            let transaction_to_add = StarknetTransaction::create_rejected(
+                Transaction::DeployAccount(Box::new(deploy_account_transaction)),
+                tx_err,
+            );
+
+            starknet.transactions.insert(&transaction_hash, transaction_to_add);
+            // Revert to previous pending state
+            starknet.state.pending_state = state_before_txn;
+        }
+    }
+
+    Ok((transaction_hash, address))
+}
+
+#[cfg(test)]
+mod tests {
+    use starknet_rs_core::types::TransactionStatus;
+    use starknet_types::contract_address::ContractAddress;
+    use starknet_types::contract_storage_key::ContractStorageKey;
+    use starknet_types::felt::Felt;
+    use starknet_types::traits::HashProducer;
+
+    use crate::account::Account;
+    use crate::constants::{self, DEVNET_DEFAULT_CHAIN_ID};
+    use crate::starknet::{predeployed, Starknet};
+    use crate::traits::{Accounted, HashIdentifiedMut, StateChanger, StateExtractor};
+    use crate::transactions::deploy_account_transaction::DeployAccountTransaction;
+    use crate::utils::test_utils::dummy_felt;
+    use crate::utils::{get_storage_var_address, load_cairo_0_contract_class};
+
+    #[test]
+    fn deploy_account_transaction_should_fail_due_to_low_balance() {
```

Comment:
I feel like these tests would have been simpler if we replaced them with e2e equivalents relying on BackgroundDevnet. Another benefit is that we would be testing the actual Devnet with actual requests. But that's just a thought, I believe the tests are valid as they are now, we don't yet have to make this replacement  (if we are even ever going to make it).

Reply:
You don't